### PR TITLE
Library implementation and go project initialization

### PR DESCRIPTION
**Pull request by bejelith**

### CODE REVIEW

File: executor/mount.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		if tokens[3] == tokens[4] && tokens[3] == "/" {
+			majorMinor := strings.Split(tokens[2], ":")
+			if len(majorMinor) != 2 {
+				return 0, 0, fmt.Errorf(
+					"the kernel may be wrong, found %d columns as maj:min:%s",
+					len(tokens),
+					line)
+			}
+			maj, _ := strconv.Atoi(majorMinor[0])
+			min, _ := strconv.Atoi(majorMinor[1])
```
Comment (rcanderson23) [inline executor/mount.go:35]: No issues ignoring errors here?

Comment (bejelith) [reply executor/mount.go:35]: I assume the kernel is always correct, in case of an error maj and minor will be 0 and not applied

File: executor/job.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	Cmd string
+	// Args is a slice of arguments to pass to Cmd
+	Args []string
+	// CPUPercent represents the quota of cpu to use for all cores. We would't assume the user knows
+	// the number of cores available so the minimum value is 1 and max is 100.
+	CPUPercent  int
+	deviceMajor int
+	deviceMinor int
+	// MemoryMB represents the quota of memory to in Megabytes, it will be applied as Memory High in the CGroup
+	MemoryMB int
```
[the remark this answers is not in the document]
Comment (bejelith) [reply executor/job.go:26]: Yeah it doesnt

File: executor/executor.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+// Get returns a process by it's ID and returns nil if ID is invalid
+func (s *Executor) Get(ID uint64) *ProcInfo {
+	s.mutex.Lock()
+	p, found := s.jobs[ID]
+	s.mutex.Unlock()
+	if !found {
+		return nil
+	}
+	return &ProcInfo{
```
Comment (tcsc) [inline executor/executor.go:69]: Is there a race here? Can other threads manipulate `p` in any way that will break stuff?

File: executor/executor.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		done:      make(chan struct{}),
+	}
+	return s, nil
+}
+
+// Get returns a process by it's ID and returns nil if ID is invalid
+func (s *Executor) Get(ID uint64) *ProcInfo {
+	s.mutex.Lock()
+	p, found := s.jobs[ID]
+	s.mutex.Unlock()
```
Comment (tcsc) [inline executor/executor.go:65]:
```suggestion
	s.mutex.RUnlock()
```

File: executor/job_linux.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		if err = os.WriteFile(cgroupPath+"/memory.max", memFile, 0664); err != nil {
+			return
+		}
+		if err = os.WriteFile(cgroupPath+"/memory.high", memFile, 0664); err != nil {
+			return
+		}
+	}
+
+	// Minimum acceptable value is 1
+	if c.deviceMajor > 1 && c.ReadBPS > 1 && c.WriteBPS > 1 {
```
Comment (tcsc) [inline executor/job_linux.go:56]: Possible ~stupid~ naïve question: What if I want to have unbounded reads but bounded writes. Does that work?

File: executor/executor.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		nextID:    -1,
+		wg:        &sync.WaitGroup{},
+		done:      make(chan struct{}),
+	}
+	return s, nil
+}
+
+// Get returns a process by it's ID and returns nil if ID is invalid
+func (s *Executor) Get(ID uint64) *ProcInfo {
+	s.mutex.Lock()
```
[the remark this answers is not in the document]
Comment (bejelith) [reply executor/executor.go:63]: Ah thx for catching that

File: executor/job_linux.go (hunk 5 above)
Comment (bejelith) [reply executor/job_linux.go:56]: Does it work in what sense? Would the kernel allow me to setup the cgroup that way or is more a question if the process under is able to function?

On point of view 1: from documentation seems it's allowed, have not tested that yet.
On point of view 2: From a theoretical point of view any process would work imho unless the bounded component ist too low to make it unoperable or stuck .

File: executor/executor.go (hunk 3 above)
Comment (bejelith) [reply executor/executor.go:69]: Yes, I should have used process.Status() which is implemented as synchronized method. 
Probably my main mistake is to have the process struct in the same package as Executor tempting to access private members.

File: executor/job_linux.go (hunk 5 above)
Comment (bejelith) [reply executor/job_linux.go]: I pushed a logic that considers both parameters as independent

File: executor/executor.go (hunk 3 above)
Comment (bejelith) [reply executor/executor.go:70]: So I think i fixed it now :)

File: executor/job_linux.go (hunk 5 above)
Comment (tcsc) [reply executor/job_linux.go]: As I read this (and I may well be missing something), if `c.WriteBPS > 1` but `c.ReadBPS == 0`, no limit of any kind will be set.

File: executor/job_linux.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		memFile := []byte(fmt.Sprintf("%d\n", c.MemoryMB*1024*1024))
+		if err = os.WriteFile(cgroupPath+"/memory.max", memFile, 0664); err != nil {
+			return
+		}
+		if err = os.WriteFile(cgroupPath+"/memory.high", memFile, 0664); err != nil {
+			return
+		}
+	}
+
+	// Minimum acceptable value is 1
```
Comment (tcsc) [inline executor/job_linux.go:58]: I think comment and code disagree, here. If Im reading it right, code says minimum acceptable value is 2 (i.e. `> 1`). 

Also, where does this requirement come from - cGroups doc or this implementation?

File: executor/job_linux_test.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+package executor
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+)
+
+func TestCGroupCreation(t *testing.T) {
```
Comment (tcsc) [inline executor/job_linux_test.go:12]: Yo may want to add tests for boundary conditions

File: executor/job_linux.go (hunk 7 above)
Comment (bejelith) [reply executor/job_linux.go:58]: i didnt find it documentation but i encountered an error yesterday.. i can replicate with
```
[root@ip-172-31-20-250 0]# echo "202:0 rbps=1" >io.max
-bash: echo: write error: Invalid argument
[root@ip-172-31-20-250 0]# echo "202:0 rbps=2" >io.max
[root@ip-172-31-20-250 0]#
```

File: executor/job.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+// Error returns and error if child process terminated unsuccessfully, nil otherwise
+func (p *process) Error() error {
+	return p.Status().err
+}
+
+func (p *process) Stop() {
+	if p.cancelFunction != nil {
+		p.cancelFunction()
```
Comment (sclevine) [inline executor/job.go:123]: Canceling will send SIGKILL to the job. This makes the job runner unsuitable for long-running commands that may require graceful termination. Is this limitation intentional?

File: internal/mount/mount_linux.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+// mount proc to the local namespace, need this to make the project build on mac
+func Mount() error {
+	// We are in the new namespace after Clone/Exec so mount /proc
+	return unix.Mount("", "/proc", "proc", syscall.CLOCAL, "")
+}
+
+func SetRootMountPrivate() error {
+	return syscall.Mount("", "/proc", "proc", syscall.MS_PRIVATE|syscall.MS_REMOUNT, "")
```
Comment (sclevine) [inline internal/mount/mount_linux.go:21]: Should this be `/proc`?

File: executor/job.go (hunk 9 above)
Comment (bejelith) [reply executor/job.go:123]: i have assume cancellation sends SIGTERM, will rework that.

File: internal/mount/mount_linux.go (hunk 10 above)
Comment (bejelith) [reply internal/mount/mount_linux.go:21]: it can be 'none', 'proc' or empty string from the tests i did

File: executor/job_linux_test.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+		// i'd read the original file content and verify it hasn't mutated.
+		// This is useful when the content of the file is not static, in this case is set by the OS.
+		errExpected: true,
+	}}
+
+	for _, test := range files {
+		os.Remove(test.file)
+
+		t.Run(test.name, func(t *testing.T) {
+			fd, cgroupDir, err := writeCgroup(0, test.config, basePath)
```
Comment (tcsc) [inline executor/job_linux_test.go:67]:
```suggestion
			fd, cgroupDir, err := writeCgroup(0, test.config, t.TempDir())
```

File: executor/job_linux_test.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+		config: ProcessConfig{MemoryMB: 5},
+		file:   "memory.high",
+		// This is a shortcut, for tests that do not intend to modify the file
+		// i'd read the original file content and verify it hasn't mutated.
+		// This is useful when the content of the file is not static, in this case is set by the OS.
+		errExpected: true,
+	}}
+
+	for _, test := range files {
+		os.Remove(test.file)
```
Comment (tcsc) [inline executor/job_linux_test.go:64]:
```suggestion
```
Replace with automatically cleaned-up `t.TempDir()`; see below.

File: executor/job_linux_test.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+)
+
+func TestCGroupCreation(t *testing.T) {
+	basePath, _ := os.MkdirTemp("", "*")
```
Comment (tcsc) [inline executor/job_linux_test.go:13]:
```suggestion
```
Replace with automatically cleaned-up `t.TempDir()`; see below.

File: executor/job_linux_test.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+			}
+
+			trimmed := strings.Trim(string(output), " \n")
+			if trimmed != test.output && !test.errExpected {
+				t.Fatalf("expected \"%s\" but received \"%s\"", test.output, trimmed)
+			}
+		})
+
+	}
+	t.Cleanup(func() { os.Remove(basePath) })
```
Comment (tcsc) [inline executor/job_linux_test.go:90]:
```suggestion
```

File: executor/job_linux.go [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+var maxCpuTime = uint(100000)
+
+// setupCgroup can create create only children to the current cgroup
+func (p *process) setupCgroup() (int, string, error) {
+	return writeCgroup(p.ID, p.config, CGROUP_PATH)
+}
+
+// writeCgroup builds and configures are new Cgroup for the Process (pid)
+// at the moment blindly applies the values and does not verify they are coherent with system resources
```
[the remark this answers is not in the document]
Comment (sclevine) [reply executor/job_linux.go]: Sorry, I meant, in the function below you're ignoring all of errors.

File: cmd/server/main.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
+package main
```
[the remark this answers is not in the document]
Comment (bejelith) [reply cmd/server/main.go]: oh sorry, that was an accident.

File: internal/mount/mount_linux.go (hunk 10 above)
Comment (sclevine) [reply internal/mount/mount_linux.go]: I'm curious why you are remounting `/proc` on the host as private, instead of mounting over each job's `/`, inside of the job namespace? This seems rather intrusive for the host.

(I thought that was what you attempting, given the function name, but I see this is running outside of the mount namespace in your example.)

Comment (bejelith) [reply internal/mount/mount_linux.go]: This is a shortcut so that i dont have to chroot and create a new root FS, by making "/proc" private i can remount it inside the new mount namespace without leaking in the parent

File: executor/job_linux.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (bejelith) [reply executor/job_linux.go]: i think it's ok now, can u confirm?

File: internal/mount/mount_linux.go (hunk 10 above)
Comment (bejelith) [reply internal/mount/mount_linux.go]: @sclevine thx for the hint :)
I have been doing a bit more digging and I found how i can avoid remounting in the root namespace with [f2ea529](https://github.com/bejelith/goteleport/pull/2/commits/f2ea52948e0dbefc8c87925a3c5e373763ac9109)

File: internal/mount/mount_linux.go [hunk 17]
```diff
@@ -11,16 +11,15 @@ import (
 	"golang.org/x/sys/unix"
 )
 
-// mount proc to the local namespace, need this to make the project build on mac
+// mount /proc as private and then remount it as CLOCAL inside the new namespace
 func Mount() error {
+	if err := syscall.Mount("proc", "/proc", "proc", syscall.MS_PRIVATE, ""); err != nil {
```
Comment (sclevine) [inline internal/mount/mount_linux.go:16]: Very close, but might as well mount all of `/`, so the mount namespace is truly isolating. (like `mount --make-private /`)

Comment (bejelith) [reply internal/mount/mount_linux.go]: well thx for pushing me, now it hides all mount points not just proc

**Review by sclevine — APPROVED** [review]

**Comment by bejelith** [comment]
thx for the stamp!
